docs/source/caching/programmatic-cache-keys.mdx

```diff
@@ -0,0 +1,207 @@
+---
+title: Programmatic Cache Keys
+---
+
+When [declarative cache IDs](./cache-key-resolution) don't fit your use case, you can programmatically generate cache IDs for object types in your normalized cache.
+
+To configure cache keys programmatically, you can alter the [`SchemaConfiguration.cacheKeyInfo(for type:object:)`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/schemaconfiguration/cachekeyinfo(for:object:)) method to return a `CacheKeyInfo` object.
+
+<Note>
+
+Programmatic cache key usage will take precedence over declarative cache keys. If you return a `CacheKeyInfo` for an object, it will be used to compute that object's cache key and any relevant `@typePolicy` will be ignored. Returning `nil` from this function will still fall back to using declarative cache keys.
+
+</Note>
+
+## `CacheKeyInfo`
+
+The information needed to construct a **cache key** is represented by a [`CacheKeyInfo`](https://www.apollographql.com/docs/ios/docc/documentation/apolloapi/cachekeyinfo) value. This `struct` consists of two properties you can provide to determine how a cache key is computed:
+
+1. `let uniqueKeyGroup: String?`
+
+    An optional **group identifier** for a set of objects that should be grouped together in the normalized cache. This is used as the first component of the **cache key**.
+
+    <Caution>
+
+    To ensure cache key group uniqueness, all objects with the same `uniqueKeyGroup` must have unique `id`s across all types.
+
+    </Caution >
+
+    To prevent cache key collisions, cache keys will always have a **group identifier** component. When the `uniqueKeyGroup` is `nil` (the default value), the [`__typename`](https://spec.graphql.org/draft/#sec-Type-Name-Introspection) of the response object is used as the **group identifier**.
+
+    If multiple distinct types can be grouped together in the cache, the `CacheKeyInfo` for each `Object` should have the same `uniqueKeyGroup`.
+
+    <Tip>
+
+    By grouping objects together, their **keys** in the normalized cache will have the same prefix. This allows you to search for cached objects in the same group by their cache `id`. To learn more, read about [direct cache access](./cache-transactions)
+
+    </Tip>
+
+2. `let id: String`
+
+    The unique **cache ID** representing the object. This is used as the second component of the **cache key**.
+
+    <Caution>
+
+    To ensure cache ID uniqueness, the `id` must be deterministic and unique for all objects with the same **group identifier** (`__typename` or `uniqueKeyGroup`).
+
+    That is, a response object for the same entity will always have the same key in the cache.  Different objects, even with the same group identifier, will always have distinct keys.
+
+    </Caution>
+
+The normalized cache constructs cache keys with the format:
+
+`"${GroupIdentifier}:${CacheID}"`
+
+Given the following a `CacheKeyInfo`:
+```swift showLineNumbers=false
+CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
```

Comment:
Looks like the formatting is off here:

```suggestion

```swift showLineNumbers=false
CacheKeyInfo(id: "123", uniqueKeyGroup: "Animal")
```

```